vm/systemSmartContracts/governance.go

```diff
@@ -236,230 +229,470 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+
+	lenVoteSet := len(currentVoteSet.VoteItems)
+	if lenVoteSet > 0 {
+		lastVote := currentVoteSet.VoteItems[lenVoteSet-1]
+		if lastVote.Value != voteOption {
+			g.eei.AddReturnMessage("conflicting votes for same proposal")
+			return vmcommon.UserError
+		}
+	}
+
+	votePower, err := g.computeAccountLeveledPower(args.CallValue, currentVoteSet)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	currentVote := &VoteDetails{
+		Value:   voteOption,
+		Power:   votePower,
+		Balance: args.CallValue,
+	}
+
+	newVoteSet, updatedProposal, err := g.applyVote(currentVote, currentVoteSet, proposal)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = g.saveVoteSet(voterAddress, newVoteSet, updatedProposal)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = g.saveGeneralProposal(proposal.GitHubCommit, proposal)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+// claimFunds returns back the used funds for a particular proposal if they are unlocked. Accepts a single parameter:
+//  args.Arguments[0] - proposal reference
+func (g *governanceContract) claimFunds(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(big.NewInt(0)) != 0 {
+		g.eei.AddReturnMessage("invalid callValue, should be 0")
+		return vmcommon.UserError
+	}
+	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.Claim)
+	if err != nil {
+		g.eei.AddReturnMessage("not enough gas")
+		return vmcommon.OutOfGas
+	}
+
+	if len(args.Arguments) != 1 {
+		g.eei.AddReturnMessage("invalid number of arguments, expected 1")
+		return vmcommon.FunctionWrongSignature
+	}
+
+	endNonce := g.getEndNonceForProposal(args.Arguments[0])
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+
+	if endNonce > currentNonce {
+		g.eei.AddReturnMessage("your funds are still locked")
+		return vmcommon.UserError
+	}
+
+	voteKey := g.getVoteSetKeyForVoteWithFunds(args.Arguments[0], args.CallerAddr)
+	currentVoteSet, err := g.getOrCreateVoteSet(voteKey)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	if currentVoteSet.UsedBalance.Cmp(zero) <= 0 {
+		g.eei.AddReturnMessage("no funds to claim for this proposal")
+		return vmcommon.UserError
+	}
+
+	g.eei.SetStorage(voteKey, nil)
+
+	err = g.eei.Transfer(args.CallerAddr, g.governanceSCAddress, currentVoteSet.UsedBalance, nil, 0)
+	if err != nil {
+		g.eei.AddReturnMessage("transfer error on claimFunds function")
+		return vmcommon.ExecutionFailed
+	}
+
+	return vmcommon.Ok
+}
+
+// whiteListProposal will create a new proposal to white list an address
+func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
is there a way to remove a white listed address?

Reply:
not yet.